File: Find_First_and_Last_Position_of_Element_in_Sorted_Array/submission3.py
class Solution(object):

    def lower_bound(self, nums, target):

        ans = -1
        s = 0
        e = len(nums)-1

        while s<=e:
            mid = s + (e-s)//2
            if (nums[mid] >= target):
                ans = mid
                e = mid - 1
            else:
                s = mid + 1
        
        return ans

    
    def upper_bound(self, nums, target):

        ans = len(nums)
        s = 0
        e = len(nums)-1

        while s<=e:
            mid = s + (e-s)//2
            if (nums[mid] > target):
                ans = mid
                e = mid - 1
            else:
                s = mid + 1
        
        return ans


    def searchRange(self, nums, target):
        
        lower = self.lower_bound(nums, target)
        higher = self.upper_bound(nums, target)

        if lower==-1 or nums[lower]!=target:
            return [-1,-1]
        return [lower, higher-1]

File: Search_Insert_Position/submission3.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        
        int s=0, e=nums.size()-1;
        int mid;

        while (s<=e){
            mid = s + (e-s)/2;

            if (nums[mid]>=target){
                e = mid-1;
            }
            else {
                s = mid+1;
            }
        }

        return s;
    }
};